rubyPackages.iconv, v8: fix build with clang 16

reckenrode opened the pull request:
## Description of changes

Second batch of fixes for staging-next #263535.

* rubyPackages.iconv: https://hydra.nixos.org/build/239596701
* v8: https://hydra.nixos.org/build/239551487

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

reckenrode committed:
rubyPackages.iconv: fix build with clang 16

Fix incompatible function pointer conversion errors.

reckenrode committed:
v8: use clang 15 to build

The version of v8 in nixpkgs fails to build due to
`-Wenum-constexpr-conversion` errors. Since this will be made a hard
error in later versions of clang, use clang 15 to build v8.

reckenrode commented:
Setting to draft while I troubleshoot an issue I encountered while doing a rebuild.

reckenrode commented:
Marking ready after successfully rebuilding locally.

wegank commented on pkgs/top-level/all-packages.nix:
```diff
@@ -25465,7 +25465,19 @@ with pkgs;
-  v8 = darwin.apple_sdk_11_0.callPackage ../development/libraries/v8 { };
+  v8 = callPackage ../development/libraries/v8 (
+    let
+      stdenv' = if stdenv.cc.isClang && lib.versionAtLeast (lib.getVersion stdenv.cc.cc) "16"
+        then overrideCC stdenv (llvmPackages_15.clang.override {
+          inherit (stdenv.cc) libcxx;
+          extraPackages = [ stdenv.cc.libcxx.cxxabi ];
+        })
+        else stdenv;
```
I'm also needing this expression in places where `llvmPackages_*.stdenv` is used, as in #264830. Can we make something like `overrideLibcxx` in #265284 a stdenv adapter before #264091 is in, so that we can simply write the following?

```suggestion
      stdenv' = if stdenv.cc.isClang && lib.versionAtLeast (lib.getVersion stdenv.cc.cc) "16"
        then overrideLibcxx llvmPackages_15.stdenv
        else stdenv;
```

reckenrode replied:
Sure, I can refactor it into an adapter.

reckenrode replied:
There is one caveat. The libc++ 16 headers can’t be parsed by clang 14 or older. I’ll base the adapter’s implementation on #264091, so it can be used when a package needs an older version of clang to build. It can be deprecated once that PR is merged.

reckenrode committed:
stdenvAdapters: add overrideLibcxx

This was taken from #264091 to use in the interim before that PR lands
(sometime after the release of 23.11). It allows different versions of
clang to link the same libc++, allowing dependencies to be linked when
they are built with a different version of clang than the stdenv.

reckenrode commented:
@wegank I added the adapter and updated the PR to use it.

trofi commented on pkgs/stdenv/adapters.nix:
```diff
@@ -42,6 +42,50 @@ rec {
+    in
+    overrideCC stdenv (stdenv.cc.override {
+      inherit libcxx;
+      extraPackages = [ cxxabi pkgs.pkgsTargetTarget."llvmPackages_${lib.versions.major llvmLibcxxVersion}".compiler-rt ];
```
Does `pkgs.pkgsTargetTarget` work for you for any cross-cases? I tried `pkgsLLVM.asciidoc-full` and it fails the eval as:

```
nix build --no-link -f. pkgsLLVM.asciidoc-full
error:
       error: attribute 'llvmPackages_13' missing

       at pkgs/stdenv/adapters.nix:86:32:

           85|       inherit libcxx;
           86|       extraPackages = [ cxxabi pkgs.pkgsTargetTarget."llvmPackages_${lib.versions.major llvmLibcxxVersion}".compiler-rt ];
             |                                ^
           87|     });
```

It looks like `pkgsTargetTarget` is always empty for cross-compilers. On the contrast the following seems to work:

```diff
--- a/pkgs/stdenv/adapters.nix
+++ b/pkgs/stdenv/adapters.nix
@@ -83,7 +83,10 @@ rec {
     in
     overrideCC stdenv (stdenv.cc.override {
       inherit libcxx;
-      extraPackages = [ cxxabi pkgs.pkgsTargetTarget."llvmPackages_${lib.versions.major llvmLibcxxVersion}".compiler-rt ];
+      extraPackages = [
+        cxxabi
+        pkgs.buildPackages.targetPackages."llvmPackages_${lib.versions.major llvmLibcxxVersion}".compiler-rt
+      ];
     });

   # Override the setup script of stdenv.  Useful for testing new
```

trofi replied:
Proposed the change as https://github.com/NixOS/nixpkgs/pull/279463
